ports/fupy/Makefile

```diff
@@ -65,7 +70,11 @@ SRC_C = \
 
 # FIXME: Should use crt0 from newlib? However it seems to cause memory access
 # at 0x0
+ifeq ($(COPY_TO_MAIN_RAM), 1)
 SRC_S = crt0-$(CPU).S
+else
+SRC_S = crt0-$(CPU)-flash.S
```

Comment:
>In particular for upstreaming to MicroPython I suspect we'll need to cache versions of `crt0-lm32.S`

To be clear, do you want the `crt0-$(CPU).S` cached in `ports/fupy` or `ports/fupy/cached`? Because you also say this:

>If we want copies of them we should probably put them in ports/fupy/cached and check them into the port's MicroPython git (and maybe implement the "developer mode" override above).

Actually come to think of it, what was the original purpose of this rule anyway?
```
$(BUILD)/crt0-$(CPU).S: $(LIBBASE_DIRECTORY)/crt0-$(CPU).S
    cp $< $@
```

Why couldn't we have done the following to begin with?
```
$(BUILD)/crt0-$(CPU).o: $(LIBBASE_DIRECTORY)/crt0-$(CPU).S
    $(CC) -c $(CFLAGS) -o $@ $<
```

Reply:
@cr1901 Actually the correct place to cache a copy of the `crt0-$(CPU).S` file in the MicroPython source tree is probably [`ports/fupy/hw`](https://github.com/fupy/micropython/tree/master/ports/fupy/hw), a directory which already exists.

At present there appear to be *no* copies of `crt0-$(CPU).S` that are actually checked into the [FuPy MicroPython port](https://github.com/fupy/micropython/tree/master/ports/fupy) source tree -- instead it's relying on copying it from outside (ie `litex-buildenv` submodules), per user/build, which is a problem I'd eventually like to fix (to make it easier to upstream our port to MicroPython core).  Also at this point `$(BUILD)` is not the source tree, it's the object output tree, so that rule isn't actually copying in the source file to the source tree, it's putting it in the object directory.  Which, I agree, seems unnecessary except if one wants to rely on automagic GNU `make` rules for building it....

So my thinking here is:

*    For now, let's go with building straight from `$(LIBBASE_DIRECTORY)/crt0-$(CPU).S` with an explicit `Makefile` rule that picks up the right build flags and destination file, for all cases (see earlier discussion); and

*   Later, let's add a "Developer mode" option which will copy `$(LIBBASE_DIRECTORY)/crt0-$(CPU).S` to `ports/fupy/hw` if has changed, and an "End user" mode which will build from the `ports/fupy/hw/crt0-$(CPU).S` (cached) copy, so that end users can just compile MicroPython with less surrounding infrastructure (assuming they have, eg, a cross compiler).

Ewen

Reply:
@cr1901 Yes, I'd totally intended it to be a later PR, hence "Later" :-)   I've now made an issue (https://github.com/fupy/micropython/issues/52), which links to this PR for background, to capture that "Later" bit.  Feel free to commit the easy version now.  Apologies for mixing the "do now" / "we should do later" bits together here.

Ewen

Reply:
Cool, thanks :). I'm making the relevant changes to `crt0` naming now.